### PR TITLE
Re-PR add stm32h755II chip

### PR DESCRIPTION
**Pull request by jfbblue0922**
@acassis
@davids5

we made PR again.

・Add new CONFIG_STM32H7_PWR_LDO_SUPPLY macro to "Kconfig".
・Delete "stm32h7x5xx_rcc.c" and add ONFIG_STM32H7_PWR_LDO_SUPPLY macro to "stm32h7x3xx_rcc.c".

### CODE REVIEW

File: arch/arm/src/stm32h7/Kconfig [hunk 1]
```diff
@@ -284,11 +284,24 @@ config STM32H7_HAVE_SMPS
 	bool
 	default n
 
+choice
+	prompt "STM32 H7 Power Supply Selection"
+	default STM32H7_PWR_DIRECT_SMPS_SUPPLY
```
Comment (davids5) [inline arch/arm/src/stm32h7/Kconfig:289]: I do not think the default is correct.  Before the change the what branch of the https://github.com/apache/nuttx/pull/12394/files#diff-9dfe671931738d39252fc8e902e892d172ac118551d14e1a454e324edb0d0be4R889-R897 would have been selected.

We have to be super carful here not break existing boards. That had the default set BEFORE this change

**Review by davids5 — CHANGES_REQUESTED** [review]

File: arch/arm/src/stm32h7/Kconfig [hunk 2]
```diff
@@ -284,11 +284,27 @@ config STM32H7_HAVE_SMPS
...
+choice
+	prompt "STM32 H7 Power Supply Selection"
+	default STM32H7_PWR_DEFAULT_SUPPLY
+	depends on ARCH_CHIP_STM32H7
+	---help---
+            Select the power supply mode of your board
+
+config STM32H7_PWR_DEFAULT_SUPPLY
+	bool "Default"
+
```
Comment (davids5) [inline arch/arm/src/stm32h7/Kconfig:296]: Add help to all these  `config`s to describe what mode it is, based on the bits selected by `STM32 H7 Power Supply Selection`

File: arch/arm/src/stm32h7/stm32h7x3xx_rcc.c [hunk 3]
```diff
@@ -886,6 +886,12 @@ void stm32_stdclockconfig(void)
           STM32_PWR_CR3_SMPSEXTHP | STM32_PWR_CR3_SMPSLEVEL_MASK);
       regval |= STM32_PWR_CR3_SCUEN;
       putreg32(regval, STM32_PWR_CR3);
+#elif defined(CONFIG_STM32H7_PWR_LDO_SUPPLY)
+      regval = getreg32(STM32_PWR_CR3);
+      regval &= ~(STM32_PWR_CR3_BYPASS | STM32_PWR_CR3_SCUEN |
+          STM32_PWR_CR3_SMPSEXTHP | STM32_PWR_CR3_SMPSLEVEL_MASK);
+      regval |= STM32_PWR_CR3_LDOEN;
+      putreg32(regval, STM32_PWR_CR3);
 #else
```
[the remark this answers is not in the document]
Comment (jfbblue0922) [reply arch/arm/src/stm32h7/stm32h7x3xx_rcc.c:895]: I added comment.

File: arch/arm/src/stm32h7/Kconfig (hunk 1 above)
Comment (jfbblue0922) [reply arch/arm/src/stm32h7/Kconfig]: I changed default value.

File: arch/arm/src/stm32h7/stm32h7x3xx_rcc.c [hunk 4]
```diff
@@ -886,7 +886,19 @@ void stm32_stdclockconfig(void)
...
+      /* 4 else cases of power supply mode
```
Comment (davids5) [inline arch/arm/src/stm32h7/stm32h7x3xx_rcc.c:896]: There are 3 settings. 2 are covered above.

What is the "4 else cases of power supply mode" referring too. This it the third. It is the default STM32H7_PWR_DEFAULT_SUPPLY.  should there be another else with #error "No a STM32 H7 Power Supply Selection made"?

**Review by davids5 — CHANGES_REQUESTED** [review]
Please add help and comments as suggested and see if the #error make sense to add.

File: arch/arm/src/stm32h7/stm32h7x3xx_rcc.c (hunk 4 above)
Comment (jfbblue0922) [reply arch/arm/src/stm32h7/stm32h7x3xx_rcc.c:896]: @davids5
STM32H7 has six power supply modes to choose from.
We use LDO supply.
I suspect someone is using a DIRECT SMPS supply.
And I don't know what the first person used.
So we don't know what the 'else' power mode is.

'else' is considered to be when nothing is defined, but I don't know how to change 'default' from 'defining nothing' in 'Kconfig'.

What do you think we should do?

File: arch/arm/src/stm32h7/Kconfig (hunk 2 above)
Comment (jfbblue0922) [reply arch/arm/src/stm32h7/Kconfig:313]: @davids5
We tried putting other modes into Kconfig.

File: arch/arm/src/stm32h7/Kconfig [hunk 5]
```diff
@@ -284,11 +284,55 @@ config STM32H7_HAVE_SMPS
 	bool
 	default n
 
-config STM32H7_PWR_DIRECT_SMPS_SUPPLY
-	bool "Use direct SMPS supply mode"
+choice
+	prompt "STM32 H7 Power Supply Selection"
```
Comment (davids5) [inline arch/arm/src/stm32h7/Kconfig:288]: @jfbblue0922 Please build test these changes for other version of the IC. 

STM32H7_PWR_DEFAULT_SUPPLY - is not defined
STM32H7_HAVE_SMPS - will disable the choices

File: arch/arm/src/stm32h7/Kconfig [hunk 6]
```diff
@@ -284,11 +284,55 @@ config STM32H7_HAVE_SMPS
...
+	default STM32H7_PWR_DEFAULT_SUPPLY
 	depends on STM32H7_HAVE_SMPS
+	---help---
+	    The STM32H7x5 and STM32H7x7 support power supply configrations for the VCORE core domain and an external supply,
+	    by configuring the SMPS step-down converter and voltage regulator.
+	    Note:The SMPS step-down converter is not available on all packages.
+
+        Select the power supply mode of your board
+	    - ID:1 LDO supply mode used 'STM32H7_PWR_LDO_SUPPLY' macro
```
Comment (davids5) [inline arch/arm/src/stm32h7/Kconfig:297]: How does ID:x relate to the choices?

File: arch/arm/src/stm32h7/stm32h7x3xx_rcc.c [hunk 7]
```diff
@@ -886,6 +886,20 @@ void stm32_stdclockconfig(void)
...
+      regval |= STM32_PWR_CR3_LDOEN;
+      putreg32(regval, STM32_PWR_CR3);
+#elif defined(CONFIG_STM32H7_PWR_SMPS_LDO) || \
+      defined(CONFIG_STM32H7_PWR_EXT_SMPS_LDO) || \
+      defined(CONFIG_STM32H7_PWR_EXT_SMPS_BYPASS) || \
+      defined(CONFIG_STM32H7_PWR_BYPASS)
+      /* These 4 cases of power supply mode have no macros created.
+       * Selecting one of these modes will cause a compile error.
+       */
+      #error Macro of this power supply mode is not created yet.
```
Comment (davids5) [inline arch/arm/src/stm32h7/stm32h7x3xx_rcc.c:902]: What would a user be expected to do if they get this message?

**Review by davids5 — CHANGES_REQUESTED** [review]

File: arch/arm/src/stm32h7/Kconfig (hunk 5 above)
Comment (jfbblue0922) [reply arch/arm/src/stm32h7/Kconfig:288]: I removed the 'STM32H7_PWR_DEFAULT_SUPPLY' macro in Kconfig.

I tried setting the 'STM32H7_HAVE_SMPS' macro in Kconfig to other chips.
- The "STM32 H7 Power Supply Selection" menu does not appear for STM32H7x3xx
- The "STM32 H7 Power Supply Selection" menu appears for STM32H7x5xx or STM32H7x7xx

File: arch/arm/src/stm32h7/Kconfig (hunk 6 above)
Comment (jfbblue0922) [reply arch/arm/src/stm32h7/Kconfig:300]: This ID is from the data sheet (Reference manual RM0399).
Power control (PWR)
- Figure 22. System supply configurations
Figure22

- Table 33. Supply configuration control
Table33

File: arch/arm/src/stm32h7/stm32h7x3xx_rcc.c (hunk 7 above)
Comment (jfbblue0922) [reply arch/arm/src/stm32h7/stm32h7x3xx_rcc.c:902]: Since I cannot verify the operation of these modes, I hope that those who use them will add macros.

Comment (davids5) [reply arch/arm/src/stm32h7/stm32h7x3xx_rcc.c:902]: The correct thing to do is: The Kconfig should name the 6 configurations available based on the package and IC Family  and the code should set them. The default should be what it was before this PR. 

Putting a partial solution and jeopardizing breaking existing boards is totally un-acceptable.

**Review by davids5 — CHANGES_REQUESTED** [review]
The correct thing to do is: The Kconfig should name the 6 configurations available based on the package and IC Family  and the code should set them. The default should be what it was before this PR. 

Putting a partial solution and jeopardizing breaking existing boards is totally un-acceptable.

File: arch/arm/src/stm32h7/stm32h7x3xx_rcc.c (hunk 7 above)
Comment (jfbblue0922) [reply arch/arm/src/stm32h7/stm32h7x3xx_rcc.c:902]: @davids5
Before this PR, we could only use DIRECT_SMPS_SYPPLY mode or else.
And we just want to use LDO_SUPPLY mode.
Should we remove the 4 modes that no one is currently using?

And we are trying to be careful not to break the existing board.

Comment (davids5) [reply arch/arm/src/stm32h7/stm32h7x3xx_rcc.c:902]: @jfbblue0922 I will spend some time in the next couple of days reviewing the data sheets the  and push some changes to this PR.

**Comment by davids5** [comment]
@jfbblue0922 Please see https://github.com/nuttx-to-asf/incubator-nuttx/commit/c6cedbaa23161950b18caed94af54b549ef8a4b1 
It is the approach I had suggested originally. It adds the LDO mode with a chip the has SMPS. Please cherry-pick it to your branch and test it does what you needed. If it is fine then please squash it and force push.

**Comment by jfbblue0922** [comment]
@davids5

I agree with your method.
I've confirmed that this method works correctly.
Thank you for your help.

**Comment by davids5** [comment]
> @davids5
> 
> I agree with your method. I've confirmed that this method works correctly. Thank you for your help.

Great! You are very welcome. Now we just have to get CI to be happy.

Please rebase on current NuttX master and force push.

**Comment by davids5** [comment]
@jfbblue0922 - hmm Look Like CI is still not happy. @xiaoxiang781216 Any ideas what it broken?

**Comment by xiaoxiang781216** [comment]
ci break is fixed one week ago, @jfbblue0922  please rebase your patch to the latest master.

**Comment by jfbblue0922** [comment]
@davids5
@xiaoxiang781216

I rebased and pushed but nothing changed.
What should we do?

File: arch/arm/src/stm32h7/Kconfig [hunk 8]
```diff
@@ -284,11 +284,23 @@ config STM32H7_HAVE_SMPS
 	bool
 	default n
 
-config STM32H7_PWR_DIRECT_SMPS_SUPPLY
-	bool "Use direct SMPS supply mode"
+choice
+	prompt "STM32 H7 Power Supply Selection"
+	default STM32H7_PWR_DIRECT_SMPS_SUPPLY
```
Comment (xiaoxiang781216) [inline arch/arm/src/stm32h7/Kconfig:289]: where you define STM32H7_PWR_DIRECT_SMPS_SUPPLY

File: arch/arm/src/stm32h7/Kconfig [hunk 9]
```diff
@@ -284,11 +284,23 @@ config STM32H7_HAVE_SMPS
...
+	    by configuring the SMPS step-down converter and voltage regulator.
+	    Note:The SMPS step-down converter is not available on all packages.
+
+        Currenlty the only supported modes are Direct SMPS supply and LDO supply.
+
+config STM32H7_PWR_LDO_SUPPLY
+	bool "Use LDO supply mode"
 	default n
 
+endchoice #"STM32 H7 Power Supply Selection"
```
Comment (xiaoxiang781216) [inline arch/arm/src/stm32h7/Kconfig:302]:
```suggestion
endchoice

File: arch/arm/src/stm32h7/Kconfig [hunk 10]
```diff
@@ -284,10 +284,24 @@ config STM32H7_HAVE_SMPS
 	bool
 	default n
 
+choice
+	prompt "STM32 H7 Power Supply Selection"
+	default STM32H7_PWR_DIRECT_SMPS_SUPPLY
+	depends on !ARCH_BOARD_NUCLEO_H745ZI && !NUCLEOH745ZI_SMPS
```
Comment (davids5) [inline arch/arm/src/stm32h7/Kconfig:290]: A board can not be used here.

**Review by davids5 — CHANGES_REQUESTED** [review]

File: boards/Kconfig [hunk 11]
```diff
@@ -1747,6 +1747,7 @@ config ARCH_BOARD_NUCLEO_H745ZI
 	depends on ARCH_CHIP_STM32H745ZI
 	select ARCH_HAVE_LEDS
 	select ARCH_HAVE_BUTTONS
+	select HAS_PWR_DIRECT_SMPS_SUPPLY
```
Comment (xiaoxiang781216) [inline boards/Kconfig:1750]: where you define HAS_PWR_DIRECT_SMPS_SUPPLY

File: arch/arm/src/stm32h7/Kconfig [hunk 12]
```diff
@@ -284,10 +284,24 @@ config STM32H7_HAVE_SMPS
 	bool
 	default n
 
+choice
+	prompt "STM32 H7 Power Supply Selection"
+	default STM32H7_PWR_DIRECT_SMPS_SUPPLY
+	depends on STM32H7_HAVE_SMPS && !HAS_PWR_DIRECT_SMPS_SUPPLY
```
Comment (xiaoxiang781216) [inline arch/arm/src/stm32h7/Kconfig:290]: where you define HAS_PWR_DIRECT_SMPS_SUPPLY

Comment (jfbblue0922) [reply arch/arm/src/stm32h7/Kconfig:290]: @xiaoxiang781216 

Is HAS_PWR_DIRECT_SMPS_SUPPLY defined in the wrong place?
I'm not sure where to define it correctly.
If it's in the wrong place, please tell me the correct place.

Comment (xiaoxiang781216) [reply arch/arm/src/stm32h7/Kconfig]: All option must be defined through config keyword before using, but where do you define the new option HAS_PWR_DIRECT_SMPS_SUPPLY?

File: arch/arm/src/stm32h7/Kconfig (hunk 9 above)
Comment (jfbblue0922) [reply arch/arm/src/stm32h7/Kconfig]: @xiaoxiang781216 

I corrected it.

File: arch/arm/src/stm32h7/Kconfig (hunk 10 above)
Comment (jfbblue0922) [reply arch/arm/src/stm32h7/Kconfig]: @acassis 

Thank you for your advice.

File: boards/Kconfig (hunk 11 above)
Comment (jfbblue0922) [reply boards/Kconfig:1750]: @xiaoxiang781216 

I added HAS_PWR_DIRECT_SMPS_SUPPLY to stm32h7/Kconfig.
Thank you for your advice.

File: arch/arm/src/stm32h7/Kconfig (hunk 12 above)
Comment (jfbblue0922) [reply arch/arm/src/stm32h7/Kconfig:299]: I corrected it.

File: arch/arm/src/stm32h7/hardware/stm32h7x3xx_pwr.h [hunk 13]
```diff
@@ -129,16 +129,18 @@
...
 #ifdef CONFIG_STM32H7_HAVE_SMPS
-#define STM32_PWR_CR3_SDEN                  (1 << 2)  /* Bit 2: SMPS step-down converter enable */
-#define STM32_PWR_CR3_SMPSEXTHP             (1 << 3)  /* Bit 3: SMPS step-down converter external power delivery selection */
-#define STM32_PWR_CR3_SMPSLEVEL_SHIFT       (4)       /* BitS 4-5: SMPS step-down converter voltage output level selection */
-#  define STM32_PWR_CR3_SMPSLEVEL_MASK      (3 << STM32_PWR_CR3_SMPSLEVEL_SHIFT)
-#  define STM32_PWR_CR3_SMPSLEVEL_R         (0 << STM32_PWR_CR3_SMPSLEVEL_SHIFT) /* 00: */
-#  define STM32_PWR_CR3_SMPSLEVEL_1V8       (1 << STM32_PWR_CR3_SMPSLEVEL_SHIFT) /* 01 */
-#  define STM32_PWR_CR3_SMPSLEVEL_2V5       (2 << STM32_PWR_CR3_SMPSLEVEL_SHIFT) /* 10: */
-#  define STM32_PWR_CR3_SMPSLEVEL_2V5B      (3 << STM32_PWR_CR3_SMPSLEVEL_SHIFT) /* 11: */
+#  define STM32_PWR_CR3_SDEN                  (1 << 2)  /* Bit 2: SMPS step-down converter enable */
```
[the remark this answers is not in the document]
Comment (jfbblue0922) [reply arch/arm/src/stm32h7/hardware/stm32h7x3xx_pwr.h]: I corrected it.

File: arch/arm/src/stm32h7/Kconfig [hunk 14]
```diff
@@ -284,11 +284,34 @@ config STM32H7_HAVE_SMPS
 	bool
 	default n
 
+config HAS_PWR_DIRECT_SMPS_SUPPLY
```
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply arch/arm/src/stm32h7/Kconfig:287]: no change

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by jerpelea — COMMENTED** [review]
please add commit message

File: arch/arm/src/stm32h7/Kconfig (hunk 14 above)
[the remark this answers is not in the document]
Comment (jfbblue0922) [reply arch/arm/src/stm32h7/Kconfig]: @xiaoxiang781216 

I changed it from HAS_PWR_DIRECT_SMPS_SUPPLY to STM32H7_HAVE_PWR_DIRECT_SMPS_SUPPLY.

File: arch/arm/src/stm32h7/hardware/stm32h7x3xx_pwr.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (jfbblue0922) [reply arch/arm/src/stm32h7/hardware/stm32h7x3xx_pwr.h]: @xiaoxiang781216 

I've removed the spaces and aligned the values of define.

**Comment by jfbblue0922** [comment]
@jerpelea 

I changed it from HAS_PWR_DIRECT_SMPS_SUPPLY to STM32H7_HAVE_PWR_DIRECT_SMPS_SUPPLY
and I rebased onto the current NuttX master and force pushed.

**Comment by jerpelea** [comment]
> @jerpelea
> 
> I changed it from HAS_PWR_DIRECT_SMPS_SUPPLY to STM32H7_HAVE_PWR_DIRECT_SMPS_SUPPLY and I rebased onto the current NuttX master and force pushed.

if you check the patch there is no commit message
you can reuse the same message as the one it PR
git commit --amend 
type the message and save
git push -f 
this will update the current patch without the need to abort it 

Best regards
Alin

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**